### ISSUE TITLE
Document test execution lifecycle regarding user code and extensions

### ISSUE THREAD

**Issue by ptahchiev** [junit-team/junit5#149]
## Related Issues
- #386 
- #426 
## Deliverables
- [x] Document test execution lifecycle regarding user code (i.e., test classes, test methods, before methods, and after methods) and extensions in the User Guide
- [x] Add a diagram to the User Guide to visualize test lifecycle and where extensions are called.
## Original Issue Description

Hello,

I'm using JUnit 4.12 and I have a bunch of selenium tests. All I want is to have a rule that captures a screenshot when the test fails. Just like this guy here:

http://stackoverflow.com/questions/12078982/apply-rule-after-each-test-and-before-each-after-in-junit

Apparently it is not possible to do this with JUnit 4.x `Rule` because in my `@After` methods I logout the user and they see the homepage. So when tests fail I always get screenshots of homepage, rather than the page that broke the tests.

I'm reading now about Exception handling extension points here:

http://junit-team.github.io/junit5/

but to me it is still not clear if these extension points will be executed after the `@BeforeEach` and before the `@AfterEach`. Can you please clarify that?

Thanks.

**Comment by jlink** [comment]
Your assumption is correct. ExceptionHandlerExtensionPoints will be invoked first thing if a test method fails.

**Comment by marcphilipp** [comment]
Does this need better documentation?

**Comment by sbrannen** [comment]
@marcphilipp, yes, this definitely needs to be fully documented. In general, we will eventually need to document the order in which all extension points are executed with regard to each other, possible interactions, etc.

**Comment by jlink** [comment]
Before putting more effort into documenting extension points and their relative orders, we should first commit to an ordering approach (see #135) and a general approach regarding an invocation extension point (see #157).

**Comment by marcphilipp** [comment]
Team decision: Add diagram to user guide to visualize test lifecycle and where extensions are called.

**Comment by sbrannen** [comment]
Currently slated for M2.

**Comment by sbrannen** [comment]
FYI: I added _Related Issues_ and _Deliverables_ sections to this issue's description and changed the title to reflect the intended scope.

**Comment by littleclay** [comment]
I'm super interested in seeing the documentation of the order of test discovery and execution with the extension callback calls, etc. In the meantime, I do have a particular question.

We currently have some Selenium tests that we use a custom runner with. We do this to be able to create parameterized test methods while avoiding some of the issues we had with the built in JUnit Parameterized runner. Most of the functionality of the custom runner and rules we have can easily be replicated (and more cleanly, too ... congrats on the design of JUnit 5, I'm loving it so far!), we do override the following 2 methods:

```
protected Statement methodBlock( FrameworkMethod method ) { ... }
protected List<FrameworkMethod> computeTestMethods() { ... }
```

I know that most of what we did in the `methodBlock` override can be accomplished with the `TestInstancePostProcessor` but I haven't yet seen where we can affect test discovery (what we accomplish in the `computeTestMethods` override.

**Comment by mmerdes** [comment]
in progress

**Comment by sbrannen** [comment]
Ummm.... wasn't this resolved in 5.0 M4?

**Comment by marcphilipp** [comment]
From a user perspective: yes. Technically, M4 still contained the broken Keynote file that @mmerdes removed later. I changed it anyway 👍
